### DISCUSSION TITLE
BYTE order in ModBus

### DISCUSSION THREAD

**Discussion by TomasAdvance** [frangoteam/FUXA#275] [Q&A]
Hi Frangoteam,
Are they some posibility to define byte order in ModBus TCP for int32?
Thnaks for your answer.
We are connected to WAGO PLC 750-891, and if I understand the wago correctly the byte order for int32 is BADC. Is it any possibility to define byte order in fuxa scada system for int32 variable on modbus tcp?
Tomas

**Comment by unocelli** [comment]
Hi,
if you mean little-endian...yes

![image](https://user-images.githubusercontent.com/48684550/155152349-1f8a99a6-911f-47ef-ab7f-0a23d48a7295.png)

**Reply by TomasAdvance** [reply]
Hi, no I mean swapping bytes.
Example if you have bytes A B C D, with big and little endian you swap two bytes lite that D C B A. But if you want order like that C D A B?How can be done in Fuxa? Or it should be new feature.
I hope it is more clear
T

**Reply by TomasAdvance** [reply]
Hi,
Any idea how to do it?

**Reply by TomasAdvance** [reply]
Hi,
I found more explanation about what I try to explained.
Could be topic for implement following byte order for 32bit:

Float - Big Endian (ABCD) - currently implemented

Float - Little Endian (DCBA) - currently implemented

Float - Mid-Big Endian (BADC) - not implemented yet

Float - Mid-Little Endian (CDAB) - not implemented yet 

Similar is fot another data type. 

What do you thinking about? 
Thanks

**Comment by unocelli** [comment]
Hi,
the request seems a bit specific, in which case you use this Mid datatype?
I don't think that the library used [modbus-serial](https://github.com/yaacov/node-modbus-serial) support this format.

**Comment by TomasAdvance** [comment]
This specific byte order I have got with wago controller. It is incredibly think.
But now way how to change it.
Can be implemented that byte order as I mentioned above?
BR

**Reply by unocelli** [reply]
We can try...I created the modbus/wago branch, as a first test to see the applicability I added the type Float32MBE can you test it? 
Thanks

**Reply by TomasAdvance** [reply]
Hi,
I will test it of course.
Let me know when you release it. 
T

**Reply by TomasAdvance** [reply]
Hi,
Any news about this topic?

**Reply by unocelli** [reply]
I would appreciate it if you could do a test with the modbus/wago branch and the new type Float32MBE. so that I can understand that I am on the right way. I don't put changes in a release if we haven't tested them before.

**Comment by TomasAdvance** [comment]
Hi, 
I try to test with new code (I downloaded latest working version of code from git) but I can find that ne capability. Can you help me whera I can find it? 
BR
Tomáš 

Tomáš Pastrňák 
________________________________
Od: Umberto Nocelli ***@***.***>
Odesláno: středa 23. března 2022 21:02
Komu: frangoteam/FUXA
Kopie: TomasAdvance; Author
Předmět: Re: [frangoteam/FUXA] BYTE order in ModBus (Discussion #275)

I would appreciate it if you could do a test with the modbus/wago branch and the new type Float32MBE. so that I can understand that I am on the right way. I don't put changes in a release if we haven't tested them before.

**Comment by unocelli** [comment]
You have to download and install the [modbus/wago](https://github.com/frangoteam/FUXA/tree/modbus/wago) branch (not from master) and then you should be able to configure tag of type Float32MBE

**Reply by TomasAdvance** [reply]
Hi I try to test it but during installation I have got following errors? Any idea what can happen?
BR
Tomas

![Snímek z 2022-04-05 17-14-02](https://user-images.githubusercontent.com/97446916/161787251-4107b817-1624-4ea1-a2c5-ec3684a1fe94.png)
![Snímek z 2022-04-05 17-14-12](https://user-images.githubusercontent.com/97446916/161787396-fcdc52d6-5e69-487d-8b9d-71580b20f87e.png)
![Snímek z 2022-04-05 17-14-19](https://user-images.githubusercontent.com/97446916/161787445-a627f850-eea9-4217-a5ee-555fd180f849.png)

If I ignore above error and start FUXA, Modbus TCP connection not working, no data receiving.

**Reply by unocelli** [reply]
are warnings, you can ignore them. have you installed the plugin?

**Reply by TomasAdvance** [reply]
Hi, I have solved it, Python installation was missing in that computer. Now is working

**Comment by TomasAdvance** [comment]
Hi,
I have test it, but result is wrong as you can see in printscreen
First value must be around 230VAC and third value around 50Hz.
![Snímek z 2022-04-05 21-01-19](https://user-images.githubusercontent.com/97446916/161830683-1c10049d-5f27-41f0-b153-d4a0481d8b13.png)

I am able to provide remote access to this station if will be helpful.
Let me know
Tomas

**Reply by unocelli** [reply]
:confused:
can you tell me the value of the 4 bytes on the wago plc?

**Reply by TomasAdvance** [reply]
Hi, I make new print screen where you can see all about. Transformed value in MBpoll 1, raw data in bytes in MB poll 2.
![Snímek z 2022-04-05 21-40-38](https://user-images.githubusercontent.com/97446916/161836343-6e4ba45a-6582-498c-9d80-99743eb65a54.png)


For correct transformation is Little-Endian byte swap used.
![Snímek z 2022-04-05 21-34-49](https://user-images.githubusercontent.com/97446916/161835508-82e90d27-0c43-40d9-818d-e1145f1a40df.png)

**Reply by unocelli** [reply]
Thank you very much, the screenshots are very helpful, I will try again tomorrow.

**Reply by TomasAdvance** [reply]
ok,
As I said, if you want to have remote access to station, let me know I will arrange it.
Tomas

**Comment by unocelli** [comment]
Hi,
I make a change in datatypes, you have to unzip and copy the js file in \server\runtime\devices\modbus\ and restart the server
[datatypes.zip](https://github.com/frangoteam/FUXA/files/8430183/datatypes.zip)

**Reply by TomasAdvance** [reply]
Hi, 
it is working very well, you can see in print screen
![Snímek z 2022-04-06 22-17-21](https://user-images.githubusercontent.com/97446916/162063569-221fa070-0088-4f00-a5f8-60eb39bb6c30.png)

Please change only, the name Float32MBE to Float32MLE (Float - Mid-Little Endian) that is correct name of byte order.
Excelent work

**Reply by unocelli** [reply]
Great, I will change the name...there is only the float32 to swap? If no how I have to swap the other one?

**Reply by TomasAdvance** [reply]
It is valid for all 32 byte data type. Byte order will be same as we have in Float32. Most common will be Int and Uint data type.
Tomas

**Reply by unocelli** [reply]
I merged into the master branch, you have to reconfigure the tag with the new type name. Thanks a lot

**Reply by TomasAdvance** [reply]
ok, thanks as well
T

**Reply by j-e-e** [reply]
I found this thread while trying to troubleshoot a modbus connection. I was unable to retrieve the correct value in fuxa but had no problem in my modbus client application.

After reading this thread I edited modbus/datatypes.js and, for testing, altered the Float32MLE to use FloatLE interpretation (BADC) with byte swapping as follows:

Float32MLE: _gen(4, 'FloatLE', 2, 2)

Then the values started to appear correctly in fuxa.

So, we are missing a datatype in the list. And there might also be confusion about which byte ordering combination is actually the little-endian byte swap. 

Anyway, the (BADC) combination is currently missing from the list and that would solve my problem, regardless of how it is named.
Can this be added to the list of datatypes?

**Reply by unocelli** [reply]
Hi @j-e-e Thanks for the investigation

Could you open a PR adding the missing datatype (BADC) as a new option, so we can better understand your implementation and evaluate it properly?
